mkosi/config.py

```diff
@@ -1555,69 +1617,66 @@ def __init__(self) -> None:
         self.match_lookup = {m.name: m for m in self.MATCHES}
 
-    def parse_config(self, path: Path, namespace: argparse.Namespace, defaults: argparse.Namespace) -> bool:
-        extras = path.is_dir()
+    def match_config(self, path: Path, namespace: argparse.Namespace, defaults: argparse.Namespace) -> bool:
+        triggered = None
 
-        if path.is_dir():
-            path = path / "mkosi.conf"
+        if not path.exists():
+            return True
 
-        parser = configparser.ConfigParser(
-            delimiters="=",
-            comment_prefixes="#",
-            inline_comment_prefixes="#",
-            empty_lines_in_values=True,
-            interpolation=None,
-            strict=False,
-            dict_type=ConfigParserMultipleValues,
-        )
+        for _, k, v in parse_ini(path, only_sections=["Match"]):
+            trigger = v.startswith("|")
+            v = v.removeprefix("|")
+            negate = v.startswith("!")
+            v = v.removeprefix("!")
 
-        parser.optionxform = lambda optionstr: optionstr # type: ignore
+            if not v:
+                die("Match value cannot be empty")
 
-        if path.exists():
-            parser.read(path)
+            if (s := self.settings_lookup_by_name.get(k)):
+                if not (match := s.match):
+                    die(f"{k} cannot be used in [Match]")
 
-        if "Match" in parser.sections():
-            triggered = None
+                # If we encounter a setting in [Match] that has not been explicitly configured yet,
+                # we assign the default value first so that we can [Match] on default values for
+                # settings.
+                self.finalize_default(s, namespace, defaults)
 
-            for k, values in parser.items("Match"):
-                # If a match is specified multiple times, the values are returned concatenated by newlines
-                # for some reason.
-                for v in values.split("\n"):
-                    trigger = v.startswith("|")
-                    v = v.removeprefix("|")
-                    negate = v.startswith("!")
-                    v = v.removeprefix("!")
+                result = match(v, getattr(namespace, s.dest))
 
-                    if not v:
-                        die("Match value cannot be empty")
+            elif (m := self.match_lookup.get(k)):
+                result = m.match(v)
+            else:
+                die(f"{k} cannot be used in [Match]")
+
+            if negate:
+                result = not result
+            if not trigger and not result:
+                return False
+            if trigger:
+                triggered = bool(triggered) or result
 
-                    if (s := self.settings_lookup_by_name.get(k)):
-                        if not (match := s.match):
-                            die(f"{k} cannot be used in [Match]")
+        return triggered is not False
```

Comment:
Sine `triggered` with the default of `None` is falsey, the `bool` call can be omitted, since `triggered` is either `None`, `False` or `True`  and `bool(None)` will evaluate the same as `None` and for the bools themselves the calls are idempotent.

That being said, since all the the `triggered = triggered or result` is doing is oring in `result` and `triggered` will stay true if `result` is true at least once, this could be shortened to a
```python
if result:
    return True
```
and then `triggered` could be omitted as well and the return at the end could become a plain `return True`. If I didn't mangle anything up while playing around with it, this passes the tests on my end. All in all:
```python
            if result:
                return True

        return True
```
and `triggered = None` removed at the top.

Reply:
I don't think this works, we need to check if all non-triggering conditions match, we can't leave the loop early.